File: samples/go-service/pricing/pricing.go
// Package pricing applies discount rules to orders.
//
// SOLID in Go terms: small interface (Rule), new behaviour = new type,
// Calculator depends on the Rule abstraction injected by the caller.
// Money is int64 cents, never float64.
package pricing

import "errors"

// Order is an immutable order value.
type Order struct {
	SubtotalCents int64
	Tier          string
	Items         int
}

// Rule returns the discount in cents for an order.
type Rule interface {
	Discount(o Order) int64
}

// TierDiscount gives a percentage off per customer tier.
type TierDiscount struct{}

var tierPercent = map[string]int64{"gold": 10, "silver": 5}

// Discount implements Rule.
func (TierDiscount) Discount(o Order) int64 {
	return o.SubtotalCents * tierPercent[o.Tier] / 100
}

// BulkDiscount gives Percent off when Items >= MinItems.
type BulkDiscount struct {
	MinItems int
	Percent  int64
}

// Discount implements Rule.
func (b BulkDiscount) Discount(o Order) int64 {
	if o.Items < b.MinItems {
		return 0
	}
	return o.SubtotalCents * b.Percent / 100
}

// ErrNegativeSubtotal is returned for invalid orders.
var ErrNegativeSubtotal = errors.New("subtotal cannot be negative")

// Calculator sums rules and caps the total discount.
type Calculator struct {
	Rules          []Rule
	MaxDiscountPct int64
}

// Total returns the price after discounts, in cents.
func (c Calculator) Total(o Order) (int64, error) {
	if o.SubtotalCents < 0 {
		return 0, ErrNegativeSubtotal
	}
	var discount int64
	for _, r := range c.Rules {
		discount += r.Discount(o)
	}
	maxDiscount := o.SubtotalCents * c.MaxDiscountPct / 100
	return o.SubtotalCents - min(discount, maxDiscount), nil
}

File: samples/node-ts-service/src/pricing.ts
/**
 * Discount rules. SOLID: one small interface, new behaviour = new class,
 * PriceCalculator depends on the DiscountRule abstraction injected by the caller.
 * Money is integer cents, never floating-point currency.
 */
export interface Order {
  readonly subtotalCents: number;
  readonly tier: "gold" | "silver" | "none";
  readonly items: number;
}

export interface DiscountRule {
  discountCents(order: Order): number;
}

const TIER_PERCENT: Record<Order["tier"], number> = { gold: 10, silver: 5, none: 0 };

export class TierDiscount implements DiscountRule {
  discountCents(order: Order): number {
    return Math.floor((order.subtotalCents * TIER_PERCENT[order.tier]) / 100);
  }
}

export class BulkDiscount implements DiscountRule {
  constructor(
    private readonly minItems = 10,
    private readonly percent = 5,
  ) {}

  discountCents(order: Order): number {
    return order.items >= this.minItems ? Math.floor((order.subtotalCents * this.percent) / 100) : 0;
  }
}

export class PriceCalculator {
  constructor(
    private readonly rules: readonly DiscountRule[],
    private readonly maxDiscountPercent = 30,
  ) {}

  totalCents(order: Order): number {
    if (order.subtotalCents < 0) {
      throw new RangeError("subtotal cannot be negative");
    }
    const discount = this.rules.reduce((sum, rule) => sum + rule.discountCents(order), 0);
    const cap = Math.floor((order.subtotalCents * this.maxDiscountPercent) / 100);
    return order.subtotalCents - Math.min(discount, cap);
  }
}
